### PR TITLE
Remove unsafeAddress(of:)

### PR DESCRIPTION
**Pull request by swiftix**
#### What's in this pull request?

The is a patch for the upcoming Swift evolution proposal.

rdar://problem/18589289
Resolves #44566

### CODE REVIEW

**Comment by swiftix** [comment]
@gribozavr Do you mind reviewing this patch?

File: stdlib/private/SwiftReflectionTest/SwiftReflectionTest.swift [hunk 1]
```diff
@@ -296,7 +296,7 @@ internal func reflect(instanceAddress: UInt, kind: InstanceKind) {
 /// This reflects the stored properties of the immediate class.
 /// The superclass is not (yet?) visited.
 public func reflect(object: AnyObject) {
-  let address = unsafeAddress(of: object)
+  var address = Unmanaged.passUnretained(object).toOpaque()
```
Comment (gribozavr) [inline stdlib/private/SwiftReflectionTest/SwiftReflectionTest.swift]: This two pieces of code in `SwiftReflectionTest.swift` look buggy to me.  They need at least a `fixLifetime`, could you add it?  (`defer { _fixLifetime(object) }` at the beginning of the method.)

File: stdlib/private/SwiftReflectionTest/SwiftReflectionTest.swift [hunk 2]
```diff
@@ -296,7 +296,7 @@ internal func reflect(instanceAddress: UInt, kind: InstanceKind) {
...
   let addressValue = unsafeBitCast(address, to: UInt.self)
```
Comment (gribozavr) [inline stdlib/private/SwiftReflectionTest/SwiftReflectionTest.swift]: `let addressValue = UInt(bitPattern: address)`?

File: stdlib/public/core/Builtin.swift [hunk 3]
```diff
@@ -291,15 +278,20 @@ public func _onFastPath() {
...
 @inline(__always)
 internal func _usesNativeSwiftReferenceCounting(_ theClass: AnyClass) -> Bool {
 #if _runtime(_ObjC)
-  return swift_objc_class_usesNativeSwiftReferenceCounting(
-    unsafeAddress(of: theClass)
-  )
+  return swift_objc_class_usesNativeSwiftReferenceCounting(theClass)
 #else
   return true
 #endif
```
Comment (gribozavr) [inline stdlib/public/core/Builtin.swift]: Can we do it slightly differently?

```
#if _runtime(_ObjC)
@_silgen_name("swift_objc_class_usesNativeSwiftReferenceCounting")
func _usesNativeSwiftReferenceCounting(_ theClass: AnyClass) -> Bool
#else
@inline(__always)
func _usesNativeSwiftReferenceCounting(_ theClass: AnyClass) -> Bool {
  return true
}
#endif
```

What do you think?

**Comment by gribozavr** [comment]
Just small comments, LGTM otherwise!  Thank you!

File: stdlib/private/SwiftReflectionTest/SwiftReflectionTest.swift (hunk 1 above)
Comment (swiftix) [inline stdlib/private/SwiftReflectionTest/SwiftReflectionTest.swift]: Will do

File: stdlib/private/SwiftReflectionTest/SwiftReflectionTest.swift (hunk 2 above)
Comment (swiftix) [inline stdlib/private/SwiftReflectionTest/SwiftReflectionTest.swift]: Sure.

File: stdlib/public/core/Builtin.swift [hunk 4]
```diff
@@ -293,17 +280,19 @@ public func _onFastPath() {
...
+@_silgen_name("swift_objc_class_usesNativeSwiftReferenceCounting")
+func _usesNativeSwiftReferenceCounting(_ theClass: AnyClass) -> Bool
 #else
+@_versioned
+@inline(__always)
+func _usesNativeSwiftReferenceCounting(_ theClass: AnyClass) -> Bool {
   return true
-#endif
 }
+#endif
```
Comment (swiftix) [inline stdlib/public/core/Builtin.swift]: @gribozavr Is it what you meant? Specially, is it correct to put `@_version` just before the non `_ObjC` version?

Comment (gribozavr) [inline stdlib/public/core/Builtin.swift]: Yes, I think that's the right way to do it.

**Comment by gribozavr** [comment]
@swiftix LGTM, thank you!

**Comment by swiftix** [comment]
@gribozavr BTW, do we need to provide a deprecated attribute with a message like: "Unmanaged.passUnretained(x).toOpaque()" instead? Or do we just silently remove the APIs?

**Comment by gribozavr** [comment]
@swiftix Yes, it would be great if we could provide migration attributes for both `unsafeAddressOf(_:)` (Swift 2.x) and `unsafeAddress(of:)` (Swift 3.0 betas).

**Comment by swiftix** [comment]
@gribozavr OK, I'll add those to the methods and use `unreachable()` as their implementation.

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
This is formally approved by means of a swift-evolution process now.

**Comment by swiftix** [comment]
@parkera This PR is removing a `unsafeAddress(of:)` according to the SE-0127. This results in a bunch of errors when compiling `foundation`, because it uses this API in a couple of places. Could you help fixing it? The official suggestion is to use `Unmanaged.passUnretained(x).toOpaque()` instead of `unsafeAddress(of: x)`

**Comment by swiftix** [comment]
@parkera Forgot to provide the link to the [compilation errors](https://ci.swift.org/job/swift-PR-Linux/3041/consoleFull#719003390ee1a197b-acac-4b17-83cf-a53b95139a76)

**Comment by parkera** [comment]
@swiftix Put up a PR on swift-corelibs-foundation and I'll merge it.

**Comment by swiftix** [comment]
@swift-ci Please smoke test Linux.

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
@swift-ci Please smoke test Linux.

**Comment by swiftix** [comment]
Test failures seem to be unrelated.

**Comment by swiftix** [comment]
@gribozavr Can you please review for CCC? There are virtually no changes since last review. I just fixed merge conflicts and force-pushed again.

**Comment by gribozavr** [comment]
LGTM.

**Pull request merged**